enhancements/multi-arch/heterogeneous-architecture-clusters.md

```diff
@@ -0,0 +1,405 @@
+by default importing manifestlists if the image is manifestlisted.
+
+### OLM
+
+OLM would need to make sure that it deploys operators only to nodes whose architecture matches the architectures on which the operator is supported.
+Today, OLM has a subscription object which is used to tell it how to deploy an operator. It has support for the nodeselector field but this would
+not be sufficient to specify filtering on multiple architectures. To this end, it would have to support the nodeAffinity field which uses the
+nodeselectorterms to filter on multiple values.
+
+Instead of having OLM dictate the operator deployment on specific nodes, the individual operators themselves would specify which architectures
+they support through a `pod.spec.affinity.nodeAffinity` object which OLM will then provide the ability for the user to override via the Subscription object
+if necessary. This information should also be added to the operands and potentially be made configurable by the user requesting the operand.
+
+Essentially, when operator authors are declaring support for multiple architectures, that entails:
+
+- Making a declaration of that support via the CSV architecture annotation
+- Setting node affinity in the CSV deployment spec so that the operator is scheduled correctly.
+- Update the operator code such that when it creates/updates operand workloads, it sets those same nodeAffinity settings that are set on
+  the operator deployment.
+
+### Machine Config Operator
+
+No changes are needed in the MCO for Phase 1 as the machine-os-content image would be a manifest listed image and the machine config daemon would
+extract the relevant architecture's machine-os-content based on the node it runs on. In the future, architecture specific configurations should
+not be required, instead they need to be templatized and be generalized for certain items, for example kubelet system reserved memory could scale
+based on page size.
+
+### Machine API
+
+- As of today, machinesets can be mutated to change the instance type and image id and this could potentially result in inconsistency between existing
+machines and newly created machines. If the instance type and image is changed to one of a different architecture, there would be workers of
+different architectures with the same machineset definition. While this may not be a problem, it doesn't make for a good user experience. Hence,
+machineset specs will not allowed to be mutated. Anytime a change is required, a new machineset needs to be created and the old one needs to be scaled
+down.
+- Changes are needed in the autoscaler to make it aware of the node architecture when autoscaling from zero. The `kubernetes-io-arch` label needs to be
+set when building the NodeInfo object so the autoscaler can pick the correct pool to scale up for allocating nodes of a particular architecture.
+
+### Hypershift
```

Comment:
I'd present the Hypershift section in two different phases as follows:
Phase 1: Control plane is assumed to run on x86. Data plane can be heterogeneous, i.e different NodePools can have different archs.
Phase 2: Control plane has the ability to run in a multi arch management cluster data plane.

To support phase 1:
- Any consumer must set an heterogeneous payload input HostedCluster.spec.release.image.
- Any consumer must set an heterogeneous payload input or the particular targeted arch for NodePool.spec.release.image.
For automation purposes the above requirements can be satisfied by our CLI to default to populate .spec.release.image fields with heterogeneous images. That's fine, but the contract is the API.
TBD:
- We need to introduce a semantic for a consumer to express arch intent so we find the right ami/image format for that arch out of the payload. Today this [is hardcoded to x86](https://github.com/openshift/hypershift/blob/dfb981f089e5150a1661d7145d0d01d175279a3b/hypershift-operator/controllers/nodepool/nodepool_controller.go#L1238).
- Remove any hardcoded image arch, atm there's only capi/caprovider, we plan to eventually pick this from the payload which as mentioned above would come from the heterogeneous spec.release.image.

To support phase 2:
This should be get supported organically via again HostedCluster.spec.release.image

Reply:
thanks. this reflects the current plan better. i will rework this section to reflect this.

Reply:
reworked the whole section. @enxebre please let me know if it looks ok.

Reply:
thanks @Prashanth684 lgtm